File: TP06/tp06.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"os"
	"strings"
)

type Vertex struct {
	name    string
	level   int
	visited bool
}

func VertexCompare(u *Vertex, v *Vertex) bool {
	return u.level < v.level || (u.level == v.level && u.name < v.name)
}

type VertexHeap []*Vertex

func (h *VertexHeap) Append(elem *Vertex) {
	*h = append(*h, elem)
	h.upheap(len(*h) - 1)
}

func (h VertexHeap) swap(i int, j int) {
	h[i], h[j] = h[j], h[i]
}

func (h VertexHeap) upheap(j int) {
	parentIdx := (j - 1) >> 1
	if j > 0 && VertexCompare(h[j], h[parentIdx]) {
		h.swap(j, parentIdx)
		h.upheap(parentIdx)
	}
}

func (h VertexHeap) downheap(j int) {
	leftIdx := 2*j + 1
	if leftIdx >= len(h) {
		return
	}

	smallestChildIdx := leftIdx

	rightIdx := 2*j + 2
	if rightIdx < len(h) && VertexCompare(h[rightIdx], h[leftIdx]) {
		smallestChildIdx = rightIdx
	}

	if VertexCompare(h[smallestChildIdx], h[j]) {
		h.swap(j, smallestChildIdx)
		h.downheap(smallestChildIdx)
	}
}

func (h *VertexHeap) Pop() (*Vertex, error) {
	arrLen := len(*h)
	if arrLen == 0 {
		return nil, errors.New("attempting to pop an empty heap")
	}

	h.swap(0, arrLen-1)
	item := (*h)[arrLen-1]
	*h = (*h)[:arrLen-1]
	h.downheap(0)
	return item, nil
}

var inside map[*Vertex][]*Vertex = make(map[*Vertex][]*Vertex)
var outside map[*Vertex][]*Vertex = make(map[*Vertex][]*Vertex)
var vertexesMap map[string]*Vertex = make(map[string]*Vertex)
var vertexes []*Vertex = make([]*Vertex, 0)

func AddVertex(vName string) *Vertex {
	v := Vertex{name: vName}
	inside[&v] = make([]*Vertex, 0)
	outside[&v] = make([]*Vertex, 0)

	vertexes = append(vertexes, &v)
	vertexesMap[vName] = &v
	return &v
}

func AddEdge(source *Vertex, target *Vertex) {
	outside[source] = append(outside[source], target)
	inside[target] = append(inside[target], source)
}

func RemoveElement(l []*Vertex, item *Vertex) []*Vertex {
	for i, other := range l {
		if other == item {
			return append(l[:i], l[i+1:]...)
		}
	}
	return l
}

func RemoveEdge(source *Vertex, target *Vertex) {
	outside[source] = RemoveElement(outside[source], target)
	inside[target] = RemoveElement(inside[target], source)
}

func AddMatkul(matkul string, deps ...string) {
	if _, ok := vertexesMap[matkul]; ok {
		fmt.Printf("Matkul %s sudah ada\n", matkul)
		return
	}

	for _, dep := range deps {
		if _, ok := vertexesMap[dep]; !ok {
			fmt.Printf("Matkul %s tidak ditemukan\n", dep)
			return
		}
	}

	matkulV := AddVertex(matkul)
	for _, dep := range deps {
		depV := vertexesMap[dep]
		AddEdge(depV, matkulV)
	}
}

func EditMatkul(matkul string, deps ...string) {
	if _, ok := vertexesMap[matkul]; !ok {
		fmt.Printf("Matkul %s tidak ditemukan\n", matkul)
		return
	}

	for _, dep := range deps {
		if _, ok := vertexesMap[dep]; !ok {
			fmt.Printf("Matkul %s tidak ditemukan\n", dep)
			return
		}
	}

	matkulV := vertexesMap[matkul]

	var prevDeps []*Vertex = make([]*Vertex, len(inside[matkulV]))
	copy(prevDeps, inside[matkulV])
	for _, dep := range prevDeps {
		RemoveEdge(dep, matkulV)
	}

	for _, dep := range deps {
		depV := vertexesMap[dep]
		AddEdge(depV, matkulV)
	}
}

func VisitVertex(v *Vertex) {
	v.visited = true

	level := 1
	for _, adjV := range inside[v] {
		if !adjV.visited {
			VisitVertex(adjV)
		}

		if adjV.level > level {
			level = adjV.level
		}
	}
	v.level = level + 1
}

func PrintSorted() {
	leveledHeap := VertexHeap{}
	for _, k := range vertexes {
		if len(inside[k]) == 0 {
			k.visited = true
			k.level = 1
		} else {
			k.visited = false
			k.level = -1
		}
	}

	for _, v := range vertexes {
		if !v.visited {
			VisitVertex(v)
		}
		leveledHeap.Append(v)
	}

	sorted := make([]string, 0)
	totalElem := len(leveledHeap)
	for i := 0; i < totalElem; i++ {
		matkul, _ := leveledHeap.Pop()
		sorted = append(sorted, matkul.name)
	}

	fmt.Println(strings.Join(sorted, ", "))
}

func PrintVars() {
	for k, v := range inside {
		fmt.Println("[IN] ", k, "value is", v)
	}
	for k, v := range outside {
		fmt.Println("[OU] ", k, "value is", v)
	}
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		currentLine := strings.TrimSpace(scanner.Text())
		if currentLine == "EXIT" {
			break
		}

		cmdArgs := strings.Split(currentLine, " ")
		cmd := cmdArgs[0]
		args := cmdArgs[1:]
		switch cmd {
		case "ADD_MATKUL":
			AddMatkul(args[0], args[1:]...)
		case "EDIT_MATKUL":
			EditMatkul(args[0], args[1:]...)
		case "CETAK_URUTAN":
			PrintSorted()
		case "PRINT":
			PrintVars()
		default:
			fmt.Println("Perintah tidak ditemukan")
		}
	}
}

File: TP06/tp06.py
from collections import defaultdict

vertexes: dict[str, "Vertex"] = {}
inside: dict["Vertex", set["Vertex"]] = defaultdict(set)
outside: dict["Vertex", set["Vertex"]] = defaultdict(set)


class Vertex:
    __slots__ = ("name", "level", "seen")

    def __init__(self, name: str):
        self.name = name
        self.level = -1
        self.seen = False

    def __lt__(self, other: object):
        if isinstance(other, Vertex):
            return self.level < other.level or (
                self.level == other.level and self.name < other.name
            )

    def __hash__(self) -> int:
        return hash(id(self))


class VertexHeap:
    __slots__ = ("_data",)

    def __init__(self, iter: list[Vertex]):
        self._data = iter

    def __len__(self):
        return len(self._data)

    def _swap(self, i, j):
        self._data[i], self._data[j] = self._data[j], self._data[i]

    def _upheap(self, j: int):
        parent_idx = (j - 1) >> 1  # // 2
        if j > 0 and self._data[j] < self._data[parent_idx]:
            self._swap(j, parent_idx)
            self._upheap(parent_idx)

    def _downheap(self, j: int):
        left_idx = 2 * j + 1
        if left_idx >= len(self._data):
            # Left node doesnt exist
            return

        smallest_child = left_idx

        right_idx = 2 * j + 2
        if right_idx < len(self._data):
            if self._data[right_idx] < self._data[left_idx]:
                smallest_child = right_idx

        if self._data[smallest_child] < self._data[j]:
            self._swap(j, smallest_child)
            self._downheap(smallest_child)

    def append(self, elem: Vertex):
        self._data.append(elem)
        self._upheap(len(self._data) - 1)

    def pop(self):
        if len(self._data) == 0:
            raise IndexError("Attempting to pop an empty heap")

        self._swap(0, len(self._data) - 1)
        item = self._data.pop()
        self._downheap(0)
        return item


def add_vertex(v_name: str):
    v = Vertex(v_name)
    inside[v] = set()
    outside[v] = set()
    vertexes[v_name] = v
    return v


def add_edge(source: Vertex, target: Vertex):
    outside[source].add(target)
    inside[target].add(source)


def remove_edge(source: Vertex, target: Vertex):
    outside[source].remove(target)
    inside[target].remove(source)


def add_matkul(matkul: str, *deps: str):
    if matkul in vertexes:
        print(f"Matkul {matkul} sudah ada")
        return

    for dep in deps:
        if dep not in vertexes:
            print(f"Matkul {dep} tidak ditemukan")
            return

    v = add_vertex(matkul)
    for dep in deps:
        u = vertexes[dep]
        add_edge(u, v)


def edit_matkul(matkul: str, *deps: str):
    if matkul not in vertexes:
        print(f"Matkul {matkul} tidak ditemukan")
        return

    for dep in deps:
        if dep not in vertexes:
            print(f"Matkul {dep} tidak ditemukan")
            return

    v = vertexes[matkul]
    for u in inside[v].copy():
        remove_edge(u, v)

    for dep in deps:
        u = vertexes[dep]
        add_edge(u, v)


def print_sorted() -> None:
    leveled = VertexHeap([])
    for v in vertexes.values():
        if len(inside[v]) == 0:
            v.seen = True
            v.level = 1
            leveled.append(v)
        else:
            v.seen = False
            v.level = -1

    def visit(u: Vertex):
        u.seen = True

        max_level = 1
        for adj_v in inside[u]:
            if not adj_v.seen:
                visit(adj_v)

            max_level = max(max_level, adj_v.level)

        u.level = max_level + 1
        leveled.append(u)

    for v in vertexes.values():
        if not v.seen:
            visit(v)

    print(
        ", ".join(
            [leveled.pop().name for _ in range(len(vertexes))],
        )
    )


cmds = {
    "ADD_MATKUL": add_matkul,
    "EDIT_MATKUL": edit_matkul,
    "CETAK_URUTAN": print_sorted,
}

while True:
    cmd, *args = input().split()
    if cmd == "EXIT":
        break

    if cmd not in cmds:
        print("Perintah tidak ditemukan")
        continue

    cmds[cmd](*args)  # type: ignore
